[number input] number input doc update v11 (#3246)
carbon-design-system/carbon-website: src/pages/components/number-input/usage.mdx

<AnchorLink>Overview</AnchorLink>
<AnchorLink>Live demo</AnchorLink>
<AnchorLink>Formatting</AnchorLink>
<AnchorLink>Content</AnchorLink>
<AnchorLink>Behaviors</AnchorLink>
<AnchorLink>Related</AnchorLink>
<AnchorLink>References</AnchorLink>
<AnchorLink>Feedback</AnchorLink>

</AnchorLinks>

## Overview

> Number inputs are similar to text inputs, but used to specify a numeric value.

Reviewer: ```suggestion
Number inputs are similar to text inputs, but are used to specify only a numeric value.
```

> Number input lets users enter a numeric value and incrementally increase or

Reviewer: The beginning of the sentence is redundant with the first sentence. 
```suggestion
Number inputs incrementally increase or
```

decrease the value with a two-segment control. Number input reduces input
efforts for adjusting small values.

<Row>
<Column colLg={8}>

![number input image](images/number-input-usage-1.png)

</Column>
</Row>

### When to use

> - Enabling the user to type a number in the field

Reviewer: The first one's phrasing was confusing to me
```suggestion
- The user needs to input a numeric value
```

- Adjusting small values when increasing or decreasing them requires only a few
  clicks
- When users may not know exact values and only want to change the values that
  are relative to its current state

### When not to use

- It is best practice not to use number input if large value changes are

>   expected. For examples, when the value sets from 1 to 30, number input is not

Reviewer: ```suggestion
  expected. For example, when the value sets from 1 to 30, number input is not
```

>   a great option. In this case, use

Reviewer: Can you explain why it is not a great option? Is it because it would require many clicks to reach 30?

## Formatting

### Anatomy

<Row>
<Column colLg={8}>

![number input anatomy](images/number-input-usage-2.png)

</Column>
</Row>

1. **Label**: Text that informs the user about the content they need to enter in
   the field. It is required unless you get an approved accessibility exemption.

> 2. **Numeric value**: The value changes when users entered into the field.

Reviewer: ```suggestion
2. **Numeric value**: The value changes when user enters a value into the field or uses the subtract or add controls.
```

> 3. **Helper text** (optional, except for error and warning): Assistive text that

Reviewer: Error and warning messages aren't helper text but they do replace helper text when they appear.

Author: Oh, I see. I did the same for progress bar. Do you think it's best to call them out separately from the helper text? Or we can just put it in the same line with Helper text but explain that they replace the helper text?

Reviewer: I think putting them in the same line is fine.

   can provide additional aid or context to the user. Often used to explain the
   correct data format.
4. **Field**: The container in which a user enters data.
5. **Subtract icon**: This icon triggers decremental values.
6. **Add icon**: This icon triggers incremental values.
7. **Status icon**: Indicates the state of the number input, either error or
   warning.

### Sizing

Number input has three sizes in height: **small**, **medium**, and **large**.

> The width varies in sizes based on content, layout, and design.

Reviewer: ```suggestion
The width varies in size based on content, layout, and design.
```

| Size        | Height (px/rem) | Use case                                                                                                                                              |
| ----------- | --------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------- |
| Small (sm)  | 32 / 2          | Use when space is constricted or when placing a number input in a form that is long and complex.                                                      |
| Medium (md) | 40 / 2.5        | This is the default size and the most commonly used size. When in doubt, use the medium size.                                                         |
| Large (lg)  | 48 / 3          | Use when there is a lot of space to work with. The large size is typically used in simple forms or when a number input is placed by itself on a page. |

<Row>
<Column colLg={8}>

![number input sizes](images/number-input-usage-3.png)

</Column>
</Row>

## Content

### Main elements

#### Label

> - Always include a label so that users know what to complete the field

Reviewer: ```suggestion
- Always include a label so that users know how to complete the field
```

- Keep the label short and concise
- Use sentence-style capitalization for the label

<DoDontRow>
  <DoDont caption="Do include a label">

![Do include a label](images/number-input-usage-4.png)

  </DoDont>

>   <DoDont type="dont" caption="Do not hide label that helps user understand the context of number input">

Reviewer: ```suggestion
  <DoDont type="dont" caption="Do not hide a label that helps user understand the context of number input">
```

![Do not hide label](images/number-input-usage-5.png)

  </DoDont>
</DoDontRow>

#### Helper text

> - Indicate the input value if it has a maximum or minimum

Reviewer: ```suggestion
- Indicates the input value if it has a maximum or minimum
```

> - Helper text is an optional, but required for warning and error states

Reviewer: ```suggestion
- Helper text is optional but if present will be replaced by warning and error messages when needed
```

> ### Overflow content

Reviewer: I don't think you need this specific section unless there is an overflow behavior. Otherwise just move this content below each section in main elements.

Label, numeric value, and helper text should be short and not exceed two lines.

## Behaviors

### States

The number input has a series of states for both field and controls elements:
**enabled**, **hover**, **focus**, **error**, **warning**, and **disabled**.

| State       | When to use                                                                                                                                                                                                                                                         |
| ----------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| _Enabled_   | When the number input is live but a user is not directly interacting with it. This is commonly referred to as the default or normal state of the component. An enabled number input field should contain a default value.                                           |

> | _Hover_     | When a user is hovering over the value in the number input field or the button controls.                                                                                                                                                                            |

Reviewer: ```suggestion
| _Hover_     | When a user's mouse cursor is hovering over the field or the button controls.                                                                                                                                                                            |
```

> | _Focus_     | When a user presses `Tab` to or clicks on the number input, the field or controlls become focused, indicating the user has successfully navigated to the component.                                                                                                 |

Reviewer: ```suggestion
| _Focus_     | When a user tabs to or clicks on the number input, the field or controls become focused, indicating the user has successfully navigated to the component.                                                                                                 |
```
